client/services/FeedbackGeneratorService.js

```diff
@@ -53,8 +54,52 @@ tie.factory('FeedbackGeneratorService', [
       }
     };
 
+    var _getBuggyOutputTestFeedback = function (
+      failingTest, codeEvalResult) {
+      var buggyMessages = failingTest.getMessages();
+      var feedback = FeedbackObjectFactory.create(false);
+      var lastSnapshot = (
+        TranscriptService.getTranscript().getPreviousSnapshot());
+      if (lastSnapshot !== null) {
+        // This section makes sure to provide a new hint
+        // if the student gets stuck on the same bug by checking
+        // that they've submitted new code with the same error.
+        var previousFeedback = lastSnapshot.getFeedback();
+        var previousHintIndex = previousFeedback.getHintIndex();
+        if (previousHintIndex !== -1 && 
+          previousHintIndex < buggyMessages.length) {
+          var previousMessages = previousFeedback.getParagraphs();
+          // This could cause a problem if two different buggy outputs
+          // have the exact same hint, but that shouldn't be allowed.
+          if (previousMessages[0].getContent() == 
+            buggyMessages[previousHintIndex]) {
+            var previousCode = (
+              lastSnapshot.getCodeEvalResult().getCode());
+            if (previousCode === codeEvalResult.getCode()) {
```

Comment:
Can we just set the variable hintIndex here, and leave the construction of the feedback to outside the 'if' branch? That is: if the previous code is the same, set hintIndex to previousHintIndex, otherwise set it to 0 or previousHintIndex + 1. Construct the feedback object after that.

(Aim is to reduce some of the branching. It feels to me like a lot of this logic is just to figure out the correct hint index, and the actual construction of the feedback is orthogonal to that. Ditto for the "0" case later below.)